### DISCUSSION TITLE
help with grow workflow, and exchanges vs. cooperative_tradeoff

### DISCUSSION THREAD

**Discussion by vrmarcelino** [micom-dev/micom#25] [Q&A] [answered]
Hello!

Thanks for releasing MICOM and the tutorials.
I am trying to identify the metabolic exchanges within a community of microorganisms.

I am having trouble using the ‘grow’ workflow, but I managed to run ‘Community’ and get the com.cooperative_tradeoff. This might be a dumb question but what is the difference between the fluxes from ‘com.cooperative_tradeoff‘ and the ‘exchanges‘ we get with the grow workflow?

I have generated GEMs with CarveMe and I’m using CPLEX.
Here is my code so far:

```
# paths:
model_path = "1_interm_files"

### import MAGs table containing genome-scale model paths
mag_tb = pd.read_csv("MAGs_table.csv")

##############################
# BUILD COMMUNITY MODELS
com = Community(mag_tb)

# save this community to file (>> Can I save one pickle file per GEM?)
com.to_pickle("1_interm_files/community.pickle")


##############################
# GROWTH RATES AND FLUXES
sol = com.cooperative_tradeoff(fraction=1.0, fluxes=True, pfba=True)
wanted_fluxes = sol.fluxes

#############################
# SIMULATE GROWTH -- here is where it fails - what should the "file" be here? the .xml models
# or the community.pickle generated above? I used the same file (community.pickle) for all samples here, 
# but I suspect that this is not right.
# also - do we need relative abundances or absolute abundances (generated with metagenome data) are fine?

manifest = pd.read_csv("MAGs_manifest.csv")
manifest.head()
sample_id      species  abundance              file
0  S101C3763  random_sp_1   63740.97  community.pickle
1   S103C477  random_sp_2   56361.54  community.pickle
2   S113C133  random_sp_3   13370.47  community.pickle
3    S122C84  random_sp_4    4623.59  community.pickle
4    S132C59  random_sp_5    3604.97  community.pickle


# using the western diet as a media:
medium = load_qiime_medium("0_diet/western_diet_gut.qza")

res = grow(manifest, model_path, medium=medium, tradeoff=0.5) 
# this fails with the error:
No ID from the medium could be found in the exchange reactions. This means you probably have mismatched IDs...

```

Thanks a lot in advance!!

**Comment by cdiener** [comment, marked as answer]
Hi, there a several questions so let's get to them one by one :smile:

> This might be a dumb question but what is the difference between the fluxes from ‘com.cooperative_tradeoff‘ and the ‘exchanges‘ we get with the grow workflow?

Not a dumb question at all. The particular call that you are using is the same as using `grow(..., method="pFBA")`. By default the grow workflow uses a "cheaper" strategy where it minimizes only the global import fluxes and then gets the solution. This is slightly faster but also a bit less stable than pFBA for very large models (100+ taxa). But, yes in general it is pretty close to `Community.cooperative_tradeoff`. The major difference is that `grow` is built to simulate several samples in parallel whereas `Community.cooperative_tradeoff` only simulates a single sample.

> I am having trouble using the ‘grow’ workflow...

So there are two things that you would need to adjust for the grow workflow. 
1. The manifest is a list of *community* models. So it should contain one row for each sample. It's the format shown in cell [3] at https://micom-dev.github.io/micom/high_level.html#Building-community-models. So it would have one row per sample and would look something like that:
```bash
In [2]: import pandas as pd
   ...: 
   ...: pd.DataFrame({"sample_id": "sample_1", "file": "community.pickle"}, index=[0])
Out[2]: 
  sample_id              file
0  sample_1  community.pickle
```

2. The actual error you get is due to something completely different. The diets need to use the same ID system as the models. This is guaranteed if you use a medium matched to the model database. AGORA does not use the same names as CARVEME/BIGG so those can not be exchanged. MICOM is warning here because the metabolite names in the medium do not match your models. If your model was reconstructed using CARVEME we do provide [a diet specific for CARVEME models](https://github.com/micom-dev/media/raw/main/media/western_diet_gut_carveme.qza) in the [MICOM media repository](https://github.com/micom-dev/media). Hopefully that one works better.

**Comment by vrmarcelino** [comment]
Hi!

Thanks a lot for the quick and complete response, and for pointing to the carveme diet database.

It didn't work initially, as the reactions in the GEMs and the media had slightly different names.
Replacing "_m" with "_e_m" in the media file did the trick though:

`medium = medium.replace(regex=r'_m$', value='_e_m')`

And now the community grows! =D

Thanks again!
Vanessa

**Reply by cdiener** [reply]
Good catch. This is related to #9. I should fix this in the codebase to catch the CARVEME naming scheme for compartments.
